### ISSUE TITLE
L2G Training and Predicting are staggeringly slow

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#3779]
**Describe the bug**
Significant performance degradation observed in both L2G training and prediction phases. Model file size has also doubled between releases 24.12 and 25.03.

**Observed behaviour**
1. Training time increased from ~15 minutes to ~1 hour
2. Model file size increased from [2.6MB ](https://huggingface.co/opentargets/locus_to_gene/tree/66657530b74d25b0639740d014d85b8dbe94efc5)to [6.4MB](https://huggingface.co/opentargets/locus_to_gene/tree/main)
3. Prediction+SHAP computation time for 1M rows increased from 30 minutes to ~5 hours
4. No change in hyperparameters used during training between versions

**Expected behaviour**
- Training time should be closer to the previous ~15 minute benchmark
- Prediction and SHAP computation time should remain around 30 minutes for 1M rows

**Additional context**
Potential root causes under investigation:
- Feature complexity: a change in the features could make the model more complex. Having to extract more complex patterns might result in a larger structure.
- Dependencies: Minor scikit-learn release changes since December. I don't expect this is the issue.
- Serialization: `skops`'s December release might have affected model storage efficiency

**Next steps**: to analyse the topologies of the decision trees between the 24.12 and 25.03 models. There might be obvious structural changes that explain the loss in performance.

**Comment by ireneisdoomed** [comment]
Bare in mind that the associations datasets are wrongly located and are going to be moved to `outputs`:
```
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_by_datasource_direct/
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_by_datasource_indirect/
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_by_datatype_direct/
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_by_datatype_indirect/
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_by_overall_indirect/
gs://open-targets-pre-data-releases/2503-testrun-1/view/association_overall_direct/
```

**Comment by ireneisdoomed** [comment]
We've identified the issue: model with params after fine tuning is much more complex.

## Topology comparison for one of the decision trees (model is an ensembl of 100 decision trees)
### Old model (default params)

![Image](https://github.com/user-attachments/assets/7def0fb4-3b91-4b08-86ff-267e3a40a955)

### New model (finetuned params)

![Image](https://github.com/user-attachments/assets/0da9c3a0-0b60-441c-8a13-54fde098d47d)


**Let's recap:** this complexity is an issue mainly in the prediction step, where calculating SHAP values is the bottleneck. Training for 1h is acceptable if the performance is good, and the predictions are fast to compute. The training step however is also affected since we extract importances for the whole training set to evaluate the model.

**Next step:** to test an approximation of SHAP values where the order of the feature to assert its importance in the coalition is not taken into account. Docs:
```
approximate : bool
            Run fast, but only roughly approximate the Tree SHAP values. This runs a method
            previously proposed by Saabas which only considers a single feature ordering. Take care
            since this does not have the consistency guarantees of Shapley values and places too
            much weight on lower splits in the tree.
```
Extracting performant importances that are in line with our benchmarks would solve the issue.

**Comment by ireneisdoomed** [comment]
## Evaluation of SHAP values approximation

Approximating SHAP values massively increases the runtime for both the train and the predict step.
- Prediction: from ~5h to 13 minutes.
- Training: from 1h13min to 10 minutes.

This is really good, however estimations are not accurate enough to explain a prediction. We will:
- base the model feature importance logging in a sample of points from the training set, to speed up training
- run the predictions with the original method for 25.03, and implement a multi-threaded way of extracting SHAPs 

### Training

Graph for feature importances for each prediction in the gold standard:

<img width="916" alt="Image" src="https://github.com/user-attachments/assets/ea3e8c96-e6d3-4e8e-b34b-614c1aca4563" />

The approximation seems to maintain the general direction of feature effects and the overall  importance of distance features. However, in comparison, the magnitude is exaggerated in many cases, meaning that the approximation tends to overestimate the feature impact.

### Predicting

Using the credible set illustrated [here](https://github.com/opentargets/issues/issues/3755#issuecomment-2642782970) as an example, with and without the approximation, we see a similar trend than in the model interpretation.

I don't have a plot for the approximated SHAPs, but you can see that:
- The approximation doesn't perform well when the model output is set to `probability´. 
- Magnitude discrepancies as we saw above. `distanceSentinelFootprintNeighbourhood` has an extremely large contribution when approximating
- Opposite directional effects between methods, very obvious in `vepMaximum`

<img width="845" alt="Image" src="https://github.com/user-attachments/assets/07ccf180-a3d9-41e5-891b-968461b98a9c" />
